Get command uses print-column extn from Openapi schema

droot opened the pull request:
**What this PR does / why we need it**:

Kubectl Get command now uses metadata 'x-kubernetes-print-column' from Openapi schema to display a resource. This is to enable richer experience for non-compiled types (like service catalog API resources) in Kubectl. This functionality is currently guarded by a boolean flag "use-openapi-print-columns". 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fixes https://github.com/kubernetes/kubectl/issues/22

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Get command uses OpenAPI schema to enhance display for a resource if run with flag 'use-openapi-print-columns'. 
An example command:
kubectl get pods --use-openapi-print-columns 
```

droot committed:
Get cmd uses print-column extn from Openapi schema

Get command now uses metadata x-kubernetes-print-columns, if present, in Openapi schema
to format output for a resource. This functionality is guarded by a boolean
flag 'use-openapi-print-columns'.

pwittrock commented on pkg/api/v1/types.go:
```diff
@@ -2568,6 +2568,7 @@ type PodStatusResult struct {
 
 // Pod is a collection of containers that can run on a host. This resource is created
 // by clients and scheduled onto hosts.
+// +k8s:openapi-gen=x-kubernetes-print-columns:custom-columns=NAME:.metadata.name,RSRC:.metadata.resourceVersion
```
Good for testing, but we don't want to merge this.

pwittrock commented on pkg/kubectl/cmd/get.go:
```diff
@@ -139,6 +139,8 @@ func NewCmdGet(f cmdutil.Factory, out io.Writer, errOut io.Writer) *cobra.Comman
 // RunGet implements the generic Get command
 // TODO: convert all direct flag accessors to a struct and pass that instead of cmd
 func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args []string, options *GetOptions) error {
+	go f.OpenAPISchema(cmdutil.GetOpenAPICacheDir(cmd))
```
lets flag guard this, default off until Antoine's optimization is merged

pwittrock replied:
What  happens to this if the command exits before this is finished?  Should we defer waiting for this to finish or something?

apelisse replied:
Where does the result of this function go? Why is this running asynchronously? How do we not care about the fetched resource?

mengqiy replied:
It at least deserves a comment.

pwittrock commented on pkg/kubectl/cmd/util/factory_builder.go:
```diff
@@ -47,18 +49,34 @@ func NewBuilderFactory(clientAccessFactory ClientAccessFactory, objectMappingFac
+		// default specification to print this resource type
+		outputFormatFromOpenAPI := f.lookupDisplayColumnsFromOpenAPI(cmd, mapping)
+		if outputFormatFromOpenAPI != "" {
+			outputOpts = f.getOutputOptsFromOpenAPISpec(outputFormatFromOpenAPI)
```
should this line be wrapped into: lookupDisplayColumnsFromOpenAPI

pwittrock commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -162,6 +142,48 @@ func PrintResourceInfoForCommand(cmd *cobra.Command, info *resource.Info, f Fact
 	return printer.PrintObj(info.Object, out)
 }
 
+func extractOutputOptions(cmd *cobra.Command) *printers.OutputOptions {
```
much better.  add some documentation to the function before merge plz

pwittrock commented on pkg/printers/customcolumn.go:
```diff
@@ -153,6 +153,7 @@ type CustomColumnsPrinter struct {
 	Columns   []Column
 	Decoder   runtime.Decoder
 	NoHeaders bool
+	lastType  reflect.Type
```
what is this and why is it needed?  (add comment)

pwittrock commented:
generally looks good.  will want to add some tests.

pwittrock commented:
/assign mengqiy

droot reviewed (commented):
Thanks for the review @pwittrock . Addressed the comments. PTAL.

alexandercampbell commented on pkg/kubectl/cmd/get.go:
```diff
@@ -501,7 +513,9 @@ func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args [
 			continue
 		}
-		if err := printer.PrintObj(decodedObj, w); err != nil {
+		// TODO: understand if passing original is safe here
```
We should resolve this TODO or at least make it more specific before merging.

alexandercampbell commented on pkg/kubectl/cmd/get.go:
```diff
@@ -513,3 +527,50 @@ func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args [
+	}
+	parts := strings.SplitN(columnStr, "=", 2)
+	if len(parts) < 2 {
+		return nil
```
This should be returning `(opts *OutputOptions, ok bool)` in accordance with Golang style guides.

alexandercampbell commented on pkg/printers/interface.go:
```diff
@@ -94,3 +94,12 @@ type ErrNoDescriber struct {
+
+// OutputOptions represents resource output options which is used to generate a resource printer.
+type OutputOptions struct {
+	FmtType string // e.g. json, yaml, template, jsonpath etc.
```
Maybe this comment should point where I can find the actual list of possible values.

droot replied:
I agree. Unfortunately, these are not defined as constant types in the code. For now, I have mentioned the filename where the supported format types are listed inside a function. I would like to make a separate change for defining these valid format types as constants.

mengqiy commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -105,36 +105,19 @@ func ValidateOutputArgs(cmd *cobra.Command) error {
-	if cmd.Flags().Lookup("allow-missing-template-keys") != nil {
-		allowMissingTemplateKeys = GetFlagBool(cmd, "allow-missing-template-keys")
-	}
 	printer, generic, err := printers.GetStandardPrinter(
```
I think we can simplify the signature of `GetStandardPrinter()` by passing in `outputOptions` instead of `outputFormat, templateFile,allowMissingTemplateKeys`.

droot replied:
Done. Thanks for the suggestion.

mengqiy commented on pkg/kubectl/cmd/get.go:
```diff
@@ -501,7 +513,9 @@ func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args [
-		if err := printer.PrintObj(decodedObj, w); err != nil {
+		// TODO: understand if passing original is safe here
+		// if err := printer.PrintObj(decodedObj, w); err != nil {
+		if err := printer.PrintObj(original, w); err != nil {
```
I think we should use `decodedObj`, original may be an unversioned object.
Is there any issue when using `decodedObj`?

droot reviewed (commented):
Addressed comments and also updated the PR. PTAL.

droot reviewed (commented):
Addressed comments and updated the PR with additional comments. @pwittrock @mengqiy PTAL.

droot commented:
/assign @fabianofranz

mengqiy commented:
@k8s-bot ok to test

pwittrock commented on pkg/kubectl/cmd/get.go:
```diff
@@ -513,3 +537,50 @@ func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args [
 }
+
+func cmdSpecifiesOutputFmt(cmd *cobra.Command) bool {
+	return cmd.Flags().Lookup("output") != nil && cmdutil.GetFlagString(cmd, "output") != ""
```
why do we need to check Lookup?

droot replied:
cmdutil.GetFlagString(...) does a FATAL if the given flag is not defined. "output" flag is defined for the Get command, so its a bit more defensive check in this case. So can be removed, but doesn't hurt also.

pwittrock commented:
Looks generally correct, but not that familiar with generic printers

mengqiy commented on pkg/kubectl/cmd/get.go:
```diff
@@ -513,3 +537,50 @@ func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args [
+	return &printers.OutputOptions{
+		FmtType:          parts[0],
+		FmtArg:           parts[1],
+		AllowMissingKeys: true, // check if this is correct default behavior
```
>--allow-missing-template-keys=true: If true, ignore any errors in templates when a field or map key is missing in the template. Only applies to golang and jsonpath output formats.

I think we can remove the comment.

droot replied:
Done. PTAL

mengqiy commented on pkg/kubectl/cmd/util/openapi/openapi.go:
```diff
@@ -194,7 +194,7 @@ func (o *Resources) parseDefinition(name string, s spec.Schema) Kind {
 	}
 	if err != nil {
-		glog.Warning(err)
+		glog.V(2).Info(err)
```
Why change this?

droot replied:
During testing I discovered for some types, GVK metadata is not defined, so emit this error on stdout which is bad. So didn't have an option but to bump it to higher verbosity.

mengqiy commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -149,19 +129,63 @@ func PrinterForCommand(cmd *cobra.Command, mapper meta.RESTMapper, typer runtime
+	flags := cmd.Flags()
+
+	var outputFormat string
+	if flags.Lookup("output") != nil {
```
Is `Lookup` necessary?
I think if we have flag `output`, lookup it should always be non-nil, even unspecified.

droot replied:
Yes, I was being more defensive here, so taking it out.

mengqiy commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -149,19 +129,63 @@ func PrinterForCommand(cmd *cobra.Command, mapper meta.RESTMapper, typer runtime
+	// templates are logically optional for specifying a format.
+	// TODO once https://github.com/kubernetes/kubernetes/issues/12668 is fixed, this should fall back to GetFlagString
+	var templateFile string
+	if flags.Lookup("template") != nil {
```
Same question here.

droot replied:
I did a refactor of this code, so kept this code as is to avoid any surprises :)

droot reviewed (commented):
Thanks for the review @pwittrock @mengqiy . Addressed the comments. PTAL.

droot commented:
@k8s-bot ok to test

droot commented:
@k8s-bot pull-kubernetes-verify test this

droot commented:
/sig cli

fabianofranz commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

fabianofranz commented:
/lgtm
/approve

droot commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

droot commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

droot commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this

droot commented:
/retest

mengqiy commented:
/retest

wojtek-t commented:
/approve

wojtek-t commented:
/approve no-issue

smarterclayton commented:
Um, given how this overlaps with server side get, I'm not sure I approve this.

smarterclayton commented:
I do not want to add this to the client in 1.7.  I want this to be on the server side.  If we add this to the client, we'll just remove it in 1.8.  What's the point of that?

smarterclayton commented:
How did this get into the queue without the discussion resolution from the original proposal, saying that this is moving to the server.

smarterclayton commented:
My primary concern is that this is moving to the server in 1.8, so this flag would be removed and no one would every specify it.  So given that, if the desire to get something in for service catalog is very high, this needs to be an `--experimental-*` flag, and we need to say that the flag will be removed in 1.8.

droot commented:
@smarterclayton I have updated the PR and renamed the flag to '--experimental-use-openapi-print-columns'. Can you pl. take a look and remove "do-not-merge" label. Lets re-evaluate if we want to deprecate or remove this functionality from kubectl when we move this logic to server side in 1.8.

pwittrock commented:
> How did this get into the queue without the discussion resolution from the original proposal, saying that this is moving to the server.

@smarterclayton Please refer to the email thread "get consensus on displaying resources - sig-apimachinery / sig-cli" (sent April 17).  There you requested that we get together quickly over VC and resolve the discussion since it had been languishing.  The consensus was that while there was some overlap in the 2 solutions, they were not mutually exclusive.

One of the artifacts from that discussion was - should clients be able to print an object they have without making a request to the server.  e.g. does a Java client that wants to print an object it already fetch need to make a second request to the server to display the object?  Also how can clients display object configuration for resources defined locally or deleted from the server?  We agreed that it would be useful for clients to have some notion for how to print resources without a server roundtrip.

pwittrock commented:
Making the flag experimental is good idea and we should do that.  I would also make it deprecated so it is hidden.

smarterclayton commented:
Ok, I vaguely recall us saying:

1. this would be best effort (which means the openapi spec doesn't fit under our normal rules of backwards compatibility)
2. we'd fail gracefully if we can't fetch the columns

I had assumed that when the rest of the server side get plumbing went in, there would be no need for the flag when offline (since we would simply determine that the object was unrecognized, make a best effort pass to get the open api spec, and then print).

The rename to "experimental" satisfies my concerns about the public API shape, objection withdrawn with that change.  For 1.8, I'd expect for us to remove the flag and do the best effort fetch alongside the server side fetch attempt.

droot commented:
Thanks @smarterclayton 
@pwittrock marked the flag deprecated as well. PTAL. I need LGTM to get this PR merged now.

droot commented:
@k8s-bot pull-kubernetes-bazel test this

pwittrock commented:
@smarterclayton SGTM.  Yes, this will be a best effort fetch and very small and fast in 1.8 (when using the zipped binary proto + cache).  Thanks.

pwittrock commented:
/lgtm

smarterclayton commented:
Why did we introduce OutputOptions instead of adding those fields to PrintOptions?  I don't see anything special about them that justifies a separate object.

If there's no reason to have it separate, I'm going to move it into print options.

droot commented:
@smarterclayton I think the main distinction is that OutputOptions is used to determine a printer type to use while PrintOptions are simply args to a given Printer to format the output.

vencelova commented:
![IMG_20211101_183930](https://github.com/user-attachments/assets/b4bd1f1d-551c-4d27-a470-1f695a01a8b4)

vencelova commented:
Haiti

vencelova commented:
No máme
